### PR TITLE
ExperimentalScanner refactoring

### PR DESCRIPTION
**Pull request by mnonnenmacher**
Several refactorings of the experimental scanner code, see the commit messages for details. This is only the first part of the refactoring, another refactoring will follow which will add handling of issues during provenance resolution and might also include a cleaner implementation of #5047.

### CODE REVIEW

File: scanner/src/main/kotlin/experimental/ExperimentalScanner.kt [hunk 1]
```diff
@@ -178,6 +171,22 @@ class ExperimentalScanner(
         return controller.getNestedScanResultsByPackage()
     }
 
+    private fun resolvePackageProvenances(controller: ScanController) {
+        log.info { "Resolving provenance for ${controller.packages.size} packages." }
+
+        val mark = TimeSource.Monotonic.markNow()
```
Comment (sschuberth) [inline scanner/src/main/kotlin/experimental/ExperimentalScanner.kt:154]: Why not `measureTime`?

File: scanner/src/main/kotlin/experimental/ExperimentalScanner.kt [hunk 2]
```diff
@@ -187,6 +176,22 @@ class ExperimentalScanner(
         log.info { "Resolved provenance for ${controller.packages.size} packages in ${mark.elapsedNow()}." }
     }
 
+    private fun resolveNestedProvenances(controller: ScanController) {
+        log.info { "Resolving nested provenances for ${controller.packages.size} packages." }
+
+        val mark = TimeSource.Monotonic.markNow()
```
Comment (sschuberth) [inline scanner/src/main/kotlin/experimental/ExperimentalScanner.kt:170]: Why not `measureTime`?

File: scanner/src/main/kotlin/experimental/ExperimentalScanner.kt [hunk 3]
```diff
@@ -309,7 +297,14 @@ class ExperimentalScanner(
...
         }
 
-    private fun getStoredResults(controller: ScanController) {
+    private fun readStoredResults(controller: ScanController) {
+        log.info {
+            "Reading stored scan results for ${controller.getPackageProvenances().size} packages with " +
+                    "${controller.getAllProvenances().size} provenances."
+        }
+
+        val mark = TimeSource.Monotonic.markNow()
```
Comment (sschuberth) [inline scanner/src/main/kotlin/experimental/ExperimentalScanner.kt]: Why not `measureTime`?

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: scanner/src/main/kotlin/experimental/ExperimentalScanner.kt [hunk 4]
```diff
@@ -170,7 +170,10 @@ class ExperimentalScanner(
             "Read stored scan results in $readDuration:"
         }
         storedResults.entries.forEach { (scanner, results) ->
-            log.info { "\t${scanner.name}: Results for ${results.size} of ${allKnownProvenances.size} provenances." }
+            log.info {
+                "\t${scanner.name}: Results for ${results.filter { it.value.isNotEmpty() }.size} of " +
```
[the remark this answers is not in the document]
Comment (mnonnenmacher) [reply scanner/src/main/kotlin/experimental/ExperimentalScanner.kt]: This has been squashed into the later refactoring as it was replaced anyway.

File: scanner/src/main/kotlin/experimental/PackageProvenanceResolver.kt [hunk 5]
```diff
@@ -82,14 +85,28 @@ class DefaultPackageProvenanceResolver(
...
                 log.info {
                     "Could not resolve $sourceCodeOrigin for ${pkg.id.toCoordinates()}: ${it.collectMessagesAsString()}"
                 }
             }
         }
 
-        return UnknownProvenance
+        val message = buildString {
+            append(
+                "Could not resolve provenance for ${pkg.id.toCoordinates()} for source code origins " +
```
[the remark this answers is not in the document]
Comment (mnonnenmacher) [reply scanner/src/main/kotlin/experimental/PackageProvenanceResolver.kt:98]: I will make a separate commit in the next PR to add single quotes to all logs messages, if that's fine with you.

File: scanner/src/funTest/kotlin/experimental/DefaultPackageProvenanceResolverFunTest.kt [hunk 6]
```diff
@@ -71,7 +73,7 @@ class DefaultPackageProvenanceResolverFunTest : WordSpec() {
                     )
                 )
 
-                resolver.resolveProvenance(pkg, listOf(SourceCodeOrigin.ARTIFACT)) shouldBe UnknownProvenance
+                shouldThrow<IOException> { resolver.resolveProvenance(pkg, listOf(SourceCodeOrigin.ARTIFACT)) }
```
[the remark this answers is not in the document]
Comment (mnonnenmacher) [reply scanner/src/funTest/kotlin/experimental/DefaultPackageProvenanceResolverFunTest.kt:76]: I think it's ok to just assert the exception, we do this in many other places as well.

File: scanner/src/main/kotlin/experimental/PackageProvenanceResolver.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (sschuberth) [reply scanner/src/main/kotlin/experimental/PackageProvenanceResolver.kt:98]: Sure!

File: scanner/src/main/kotlin/experimental/ExperimentalScanner.kt (hunk 1 above)
Comment (mnonnenmacher) [reply scanner/src/main/kotlin/experimental/ExperimentalScanner.kt:154]: To reduce nesting. I find `measureTime` is often very invasive if the block is longer than 1-2 lines because it looks like control flow but is just measuring a duration.

Comment (sschuberth) [reply scanner/src/main/kotlin/experimental/ExperimentalScanner.kt:154]: I see. On the upside, `measureTime` nicely groups *what*'s being measured, which is esp. useful for more than 1-2 lines, IMO. I'm not entirely happy about the use of something else than `measureTime` here (mostly for consistency reasons), but I also won't insist on changing it.

Comment (sschuberth) [reply scanner/src/main/kotlin/experimental/ExperimentalScanner.kt:154]: Alternatively, the caller could now measure the time for the single function call.

File: scanner/src/main/kotlin/experimental/ScanController.kt [hunk 7]
```diff
@@ -188,7 +193,7 @@ class ScanController(
     /**
      * Return true if [ScanResult]s for the complete [NestedProvenance] of the package are available.
      */
-    fun hasCompleteScanResult(scanner: PackageScannerWrapper, pkg: Package): Boolean =
+    fun hasCompleteScanResult(scanner: ScannerWrapper, pkg: Package): Boolean =
```
Comment (sschuberth) [inline scanner/src/main/kotlin/experimental/ScanController.kt:196]: Just double-checking while skimming through the code: Is this change intended?

Comment (mnonnenmacher) [reply scanner/src/main/kotlin/experimental/ScanController.kt:196]: Yes, but I have moved it to the commit that introduced the function because `ScannerWrapper` made more sense from the beginning.

**Review by sschuberth — APPROVED** [review]

**Comment by mnonnenmacher** [comment]
Merged despite tests failing with cache issues. I have manually verified all related tests.
